### ISSUE TITLE
Incorrect token validation when logged in by email

### ISSUE THREAD

**Issue by ChristophWurst** [owncloud/core#25392]
When we validate a token (either browser session or device token), we use the login name stored in the token to check the password via the user manager. The checks [here](https://github.com/owncloud/core/blob/db45607ba94b3cef509fa5e45a14459931a42255/lib/private/User/Session.php#L591-L596) will fail if one uses the email address to check the password (`getLoginName()` will return the email address then).

Note that I actually did not test this, it just came up my mind and I thought I should better report, so someone can fix it :-)

cc @PVince81 @DeepDiver1975

**Comment by PVince81** [comment]
@ChristophWurst how critical is this or how likely is this to happen ?

**Comment by ChristophWurst** [comment]
I'd say this is rather critical. If this is really the case (still not tested), browser sessions and client tokens are invalidated at their first password check as the check fails with email:password.

**Comment by PVince81** [comment]
Any ideas about steps to reproduce ?

> if one uses the email address to check the password
Who is "one" ? Is that our existing code ?

**Comment by ChristophWurst** [comment]
Oh wait it actually works with oc users. However, we might run into problems with other user backends. https://github.com/owncloud/core/blob/3731b2a006eca1e96d4087212a5e779c85a002e4/core/Controller/LoginController.php#L178-L196 *converts* the email address to the UID in case the user logs in by email. That seems to be fine for internal users. But what if the user is a LDAP user? Then the user manager won't accept the password with UID:password unless the LDAP config was changed to accept the email address too.

@PVince81 now I remember we talked about that already.

**Comment by PVince81** [comment]
Last time I checked, LDAP didn't allow me to login with email at all unless I configured LDAP to use "mail" as the login field.

**Comment by ChristophWurst** [comment]
I tested with an user called 'admin' with email 'admin@domain.tld'. Logging in works, and so does creating a token. The token dialog tells me to use 'admin' as username, as the email was converted to the user's UID. In the case of this local user, the UID is equal the username. However, we can not rely on that magic as it will probably break with other user backends.

**Comment by ChristophWurst** [comment]
> Last time I checked, LDAP didn't allow me to login with email at all unless I configured LDAP to use "mail" as the login field.

That is *exactly* the problem I'm talking about

**Comment by PVince81** [comment]
I don't think it's a problem. Whatever the user entered to login in the web UI (email or other) for LDAP will also be shown as recommended login id in the app token page, which should also be the one that was stored in the token at creation time.

**Comment by ChristophWurst** [comment]
I have to disagree. It's a coincidence that it works with local users. If you do not set the LDAP option to allow the email address to log into ownCloud, then the logic breaks as it would use the UID as login name. Can you try that? Disable email login for LDAP users, log in again by email and create a token. The token dialog should show the UID as login name, which won't work.

Allowing email login for LDAP users is just a quick fix for one specific user backend and also bypasses the login-be-email logic as the user manager directly accepts the credentials.

**Comment by ChristophWurst** [comment]
https://github.com/owncloud/core/blob/3731b2a006eca1e96d4087212a5e779c85a002e4/core/Controller/LoginController.php#L179-L184 is the evil code path, which is not executed in case the user manager accept the email as login name already.

**Comment by PVince81** [comment]
Here is what I tested with the usual docker from https://github.com/owncloud/administration/tree/master/ldap-testing

A) LDAP with defaults (no email login)

1. Logged in as "zombie1"
1. Go to personal page
1. Create app password
1. User shows as "zombie1" in the field.
1. Login as zombie1+token with Webdav: works

B) LDAP with defaults + additionally tick "LDAP / AD email address"

1. Logged in as "zombie1"
1. Go to personal page
1. Create app password
1. User shows as "zombie1" in the field.
1. Login as zombie1+token with Webdav: works

1. Logged in as "zombie1@example.org"
1. Go to personal page
1. Create app password
1. User shows as "zombie1@example.org" in the field.
1. Login as zombie1+token with Webdav: works

1. Check whether the previous token for "zombie1" still works: it does

> Disable email login for LDAP users, log in again by email and create a token

If I disable email login for LDAP I can't login by email, it doesn't let me. The email logic from `tryLogin` that does `getByEmail` never works with LDAP because it calls `checkPassword()` with the LDAP UUID + password, which is always rejected as fail.

**Comment by PVince81** [comment]
forgot to mention that for all LDAP setups I also added "mail" as attribute for email in the "Advanced" tab. This is to make sure the email appears in the users page and also it makes `getByEmail` find the correct user, but will also fail because of `checkPassword()` with the LDAP UUID.

If I didn't set "mail" attribute for email, `getByEmail` will always fail as the email is not known in ownCloud at all in this case.

**Comment by PVince81** [comment]
@ChristophWurst any update ? Did I get it wrong ? Seems to work for me.

**Comment by ChristophWurst** [comment]
Well, it depends on what one expects from the *login by email* feature. It doesn't make sense to me that one has to enable email login for LDAP users explicitly on that user backend. I would have expected it to work with any user backend out of the box. I can't judge on other external user backends as I don't know their implementations. Maybe they work with email, maybe they don't.

> If I disable email login for LDAP I can't login by email, it doesn't let me. The email logic from tryLogin that does getByEmail never works with LDAP because it calls checkPassword() with the LDAP UUID + password, which is always rejected as fail.

I'd rephrase that to "it works only for internal users if their UID and username are equal".

**Comment by PVince81** [comment]
> Well, it depends on what one expects from the login by email feature. It doesn't make sense to me that one has to enable email login for LDAP users explicitly on that user backend

Agreed... maybe it's a conflicting feature because LDAP can specify fields for login.

I think for now the most important is to find out whether there is still a critical issue here, considering that login by email is blocked when not configured in LDAP.

**Comment by PVince81** [comment]
Moving to 9.1.1 to continue discussing/investigating

**Comment by dragotin** [comment]
I changed the severity to high instead of critical.

**Comment by PVince81** [comment]
Closing as I couldn't reproduce the issue with many different tries
